### PR TITLE
Add support for supplementary views

### PR DESCRIPTION
**Pull request by dehlen**
This adds support for supplementary views in collection views. If the user sets the configureSupplementaryView closure the given implementation will be called in viewForSupplementaryView:. If the implementation is not given and the closure is nil the collection view datasource will behave just like it did before. So this is a non breaking change but it makes the datasource more powerful by providing the ability to create more complex collection view layouts.

The following code snippets shows the usage:
```swift
dataSource.configureSupplementaryView = { dataSource, collectionView, kind, indexPath in
    let separatorView = collectionView.dequeueReusableSupplementaryView(ofKind: SeparatorView.kindIdentifier, withReuseIdentifier: SeparatorView.kindIdentifier, for: indexPath)
    return separatorView
}
 ```

### CODE REVIEW

File: DataSourceKit/CollectionViewDataSource.swift [hunk 1]
```diff
@@ -8,17 +8,19 @@
 
 import UIKit
 
-public class CollectionViewDataSource<CellDeclaration>: NSObject, UICollectionViewDataSource {
+open class CollectionViewDataSource<CellDeclaration>: NSObject, UICollectionViewDataSource {
```
Comment (jdisho) [inline DataSourceKit/CollectionViewDataSource.swift]: I think this is not meant to be `overridden` or `subclassed` outside the module.

File: DataSourceKit/CollectionViewDataSource.swift [hunk 2]
```diff
@@ -8,17 +8,19 @@
...
     public var cellDeclarations = [] as [CellDeclaration]
     
     private var registeredReuseIdentifiers = [] as [String]
     private let binderFromDeclaration: (CellDeclaration) -> CellBinder
-    
+    public typealias ConfigureSupplementaryView = (CollectionViewDataSource<CellDeclaration>, UICollectionView, String, IndexPath) -> UICollectionReusableView
```
Comment (jdisho) [inline DataSourceKit/CollectionViewDataSource.swift]: `public` over `private`

File: DataSourceKit/CollectionViewDataSource.swift [hunk 3]
```diff
@@ -42,4 +44,17 @@ public class CollectionViewDataSource<CellDeclaration>: NSObject, UICollectionVi
         
         return cell
     }
+
+
+    public func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
+        return configureSupplementaryView!(self, collectionView, kind, indexPath)
```
Comment (jdisho) [inline DataSourceKit/CollectionViewDataSource.swift]: No force unwrapping.

File: DataSourceKit/CollectionViewDataSource.swift [hunk 4]
```diff
@@ -8,17 +8,19 @@
...
 
-public class CollectionViewDataSource<CellDeclaration>: NSObject, UICollectionViewDataSource {
+open class CollectionViewDataSource<CellDeclaration>: NSObject, UICollectionViewDataSource {
     public var cellDeclarations = [] as [CellDeclaration]
     
     private var registeredReuseIdentifiers = [] as [String]
     private let binderFromDeclaration: (CellDeclaration) -> CellBinder
-    
+    public typealias ConfigureSupplementaryView = (CollectionViewDataSource<CellDeclaration>, UICollectionView, String, IndexPath) -> UICollectionReusableView
+    open var configureSupplementaryView: ConfigureSupplementaryView?
```
Comment (jdisho) [inline DataSourceKit/CollectionViewDataSource.swift]: Same reason as in the first comment.

File: DataSourceKit/CollectionViewDataSource.swift [hunk 5]
```diff
@@ -42,4 +44,17 @@ public class CollectionViewDataSource<CellDeclaration>: NSObject, UICollectionVi
         
         return cell
     }
+
+
+    public func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
+        return configureSupplementaryView!(self, collectionView, kind, indexPath)
+    }
+
+    open override func responds(to aSelector: Selector!) -> Bool {
```
Comment (jdisho) [inline DataSourceKit/CollectionViewDataSource.swift]: Same reason as in the first comment.

File: DataSourceKit/CollectionViewDataSource.swift [hunk 6]
```diff
@@ -42,4 +44,17 @@ public class CollectionViewDataSource<CellDeclaration>: NSObject, UICollectionVi
...
+        if aSelector == #selector(UICollectionViewDataSource.collectionView(_:viewForSupplementaryElementOfKind:at:)) {
+            return configureSupplementaryView != nil
+        } else {
```
Comment (jdisho) [inline DataSourceKit/CollectionViewDataSource.swift]: No need for `else`. You can do 
```swift 
return super.responds(to: aSelector)
```
outside `else`.

**Comment by dehlen** [comment]
Hey, thanks for your comments. I updated the PR accordingly.

**Comment by dehlen** [comment]
Hey I just updated the PR in order to resolve the new merge conflicts. Any chance to get this merged ? What are your thoughts ?

**Comment by ishkawa** [comment]
I think supplementary view should be supported in the same way as cells. In this pull request, supplementary view is supported in traditional way. I'm concerned about difference between ways to define cells and supplementary views.

Just a thought, if we separate sections by type associated with CellBinder, we might support supplementary view in the same way as cells. I will polish this idea and write some code this weekend.

**Comment by dehlen** [comment]
Sounds great  I am happy to help if needed by reviewing or helping out with an implementation.

**Comment by dehlen** [comment]
Any update on this one ?

**Pull request closed without merging**